sycl/test-e2e/KernelCompiler/sycl_host_device_data_layout.cpp

```diff
@@ -0,0 +1,162 @@
+
+  auto *align = sycl::malloc_shared<size_t>(1, q);
+  auto *size = sycl::malloc_shared<size_t>(1, q);
+  auto *equal = sycl::malloc_shared<bool>(1, q);
+
+  auto Test = [&](auto val, auto type_str, auto init_str) {
+    using namespace std::literals::string_literals;
+
+    using type = decltype(val);
+    auto kb_exe = syclexp::build(
+        kb_src,
+        syclexp::properties{syclexp::build_options{std::vector<std::string>{
+            "-DTYPE="s + type_str, "-DINIT="s + init_str}}});
+
+    sycl::kernel krn = kb_exe.ext_oneapi_get_kernel("foo");
+    auto *host = sycl::malloc_shared<type>(1, q);
+    *host = val;
+    auto *device = sycl::malloc_shared<type>(1, q);
+
+    q.submit([&](sycl::handler &cgh) {
+       cgh.set_args(host, device, align, size, equal);
+       cgh.single_task(krn);
+     }).wait();
+    auto Equal = [](const auto &lhs, const auto &rhs) {
+      using T = std::decay_t<decltype(lhs)>;
+      if constexpr (sycl::detail::is_vec_v<T> || sycl::detail::is_marray_v<T>) {
+        if (lhs.size() != rhs.size())
+          return false;
+
+        for (size_t i = 0; i < lhs.size(); ++i)
+          if (lhs[i] != rhs[i])
+            return false;
+
+        return true;
+      } else {
+        return lhs == rhs;
+      }
+    };
+    assert(Equal(*host, *device));
+    assert(*align == alignof(type));
+    assert(*size == sizeof(type));
+    assert(*equal == true);
+    sycl::free(host, q);
+    sycl::free(device, q);
+  };
+
+#define TEST(TYPE, INIT)                                                       \
+  Test(TYPE{INIT}, EXPAND_AND_STRINGIFY(TYPE), EXPAND_AND_STRINGIFY(INIT));
+#define TEST2(TYPE, INIT0, INIT1)                                              \
+  Test(TYPE{INIT0, INIT1}, EXPAND_AND_STRINGIFY(TYPE),                         \
+       EXPAND_AND_STRINGIFY(INIT0) ", " EXPAND_AND_STRINGIFY(INIT1));
+#define TEST3(TYPE, INIT0, INIT1, INIT2)                                       \
+  Test(TYPE{INIT0, INIT1, INIT2}, EXPAND_AND_STRINGIFY(TYPE),                  \
+       EXPAND_AND_STRINGIFY(INIT0) ", " EXPAND_AND_STRINGIFY(                  \
+           INIT1) ", " EXPAND_AND_STRINGIFY(INIT2));
+
+  TEST(size_t, 0x1122334455667788)
+  TEST(char, 0x12)
+
+  TEST(int8_t, 0x12)
+  TEST(int8_t, -0x12)
+  TEST(uint8_t, 0x12)
+
+  TEST(int16_t, 0x1234)
+  TEST(int16_t, -0x1234)
+  TEST(uint16_t, 0x1234)
+
+  TEST(int32_t, 0x12345678)
+  TEST(int32_t, -0x12345678)
+  TEST(uint32_t, 0x12345678)
+
+  TEST(int64_t, 0x1122334455667788)
+  TEST(int64_t, -0x1122334455667788)
+  TEST(uint64_t, 0x1122334455667788)
+
+  TEST(size_t, 0x1122334455667788)
+  TEST(ptrdiff_t, 0x1122334455667788)
+
+  TEST(float, 42.0f)
+  TEST(double, 42.0)
+
+  TEST(sycl::half, 42.0f)
+  TEST(sycl::ext::oneapi::bfloat16, 42.0f)
+
+  TEST(sycl::range<1>, 0x1122334455667788)
+  TEST2(sycl::range<2>, 0x1122334455667788, 0x1223344556677889)
+  TEST3(sycl::range<3>, 0x1122334455667788, 0x1223344556677889,
+        0x132435465768798A)
```

Comment:
It looks like you have tested most of the SYCL types that will be "trivially copyable" in KhronosGroup/SYCL-Docs#765.  The only ones missing are `id` and `nd_range`.  Can you add those too to make it complete?

Also, it would be good to add some tests for user-defined enumerations, since those are explicitly mentioned in the extension specification.  I'd suggest testing both unscoped enumerations and scoped enumerations.